packages/deploy-aws/src/templates/seagull_pipeline.ts

```diff
@@ -102,8 +102,55 @@ export class SeagullPipeline {
       stage: this.stage,
     }
 
-    stack.addSourceStage('source', lib.getSourceConfig(stageConfigParams, 0))
-    stack.addBuildStage('build', lib.getBuildConfig(stageConfigParams, 1))
+    const sourceAction = stack.addSourceStage(
+      'source',
+      lib.getSourceConfig(stageConfigParams, 0)
+    )
+    stack.addBuildActionStage(
+      'test',
+      lib.getTestConfig(stageConfigParams, 1, sourceAction.outputArtifact)
+    )
+    const buildAction = stack.addBuildActionStage('build', {
+      ...lib.getBuildConfig(stageConfigParams, 2, sourceAction.outputArtifact),
+      outputArtifacts: {
+        'secondary-artifacts': {
+          build: {
+            files: '**/*',
+            name: 'build',
+          },
+          dist: {
+            files: ['dist/**/*'],
+            name: 'dist',
+          },
+        },
+      },
+    })
+    const deployAction = stack.addBuildActionStage('deploy', {
+      ...lib.getDeployConfig(stageConfigParams, 3, sourceAction.outputArtifact),
+      additionalInputArtifacts: [buildAction.additionalOutputArtifact('dist')],
+      outputArtifacts: {
+        'secondary-artifacts': {
+          cfurl: {
+            files: ['/tmp/cfurl.txt'],
+            name: 'cfurl',
+          },
+          deploy: {
+            files: '**/*',
+            name: 'deploy',
+          },
+        },
+      },
+    })
+    stack.addBuildActionStage('end2end-test', {
+      ...lib.getTestEnd2EndConfig(
+        stageConfigParams,
+        4,
+        sourceAction.outputArtifact
+      ),
+      additionalInputArtifacts: [
+        deployAction.additionalOutputArtifact('cfurl'),
+      ],
+    })
```

Comment:
I would prefer to extract this large block of code into another file to make the code easier to read